src/token_adapter/jwt.ts

```
@@ -1,57 +1,125 @@
+    const secretOrPublicKey: string = await this._getSecretOrPublicKey(algorithm);
+
+    const data: ITokenData = await this._decode(token, secretOrPublicKey);
+
+    return data;
   }
 
-  private _sign(payload: ITokenData, secret: string, options: any): BluebirdPromise<any> {
-    return new BluebirdPromise((resolve, reject) => {
+  private _sign(payload: ITokenData, secretOrPrivateKey: string, options: any): Promise<any> {
+    return new Promise((resolve: any, reject: any): any => {
 
-      sign(payload, secret, options, (error, jwt) => {
+      sign(payload, secretOrPrivateKey, options, (error: any, jwt: any): void => {
         if (error) {
           reject(error);
         } else {
           resolve(<string> jwt);
         }
+
         return;
       });
     });
   }
 
-  private _decode(token: string): BluebirdPromise<any> {
-    return new BluebirdPromise((resolve, reject) => {
-      if (token === undefined) {
-        resolve();
-      } else {
-        verify(token, secret, (error, decodedToken) => {
-          if (error) {
-            reject(new Error('invalid token'));
-          } else {
-            resolve(<ITokenData> decodedToken);
-          }
-        });
-      }
+  private _decode(token: string, secretOrPublicKey: string): Promise<ITokenData> {
+    return new Promise((resolve: any, reject: any): any => {
+      verify(token, secretOrPublicKey, (error: any, decodedToken: any): void => {
+        if (error) {
+          reject(new Error('invalid token'));
+        } else {
+          resolve(<ITokenData> decodedToken);
+        }
+
+        return;
+      });
     });
   }
 
+  private async _getSecretOrPrivateKey(algorithm: string): Promise<string> {
+    switch (algorithm) {
+      case 'RS256':
+        const key: string = this.config && this.config.key ? this.config.key : undefined;
+        if (key === undefined) {
+          return undefined;
```

`undefined` is not a private key. This should throw an error.